src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1372 @@
+        return self.transformer.wtes
+
+    def set_input_embeddings(self, new_embeddings):
+        # one embedding layers for each codebook
+        self.transformer.wtes = new_embeddings
+
+    def _initialize_modules(self, config):
+        # initialize a modified non causal GPT-like model
+        # note that for there is one embedding layer and one lm_head for each codebook of Encodec
+        self.transformer = nn.ModuleDict(
+            {
+                "wtes": nn.ModuleList(
+                    [nn.Embedding(config.input_vocab_size, config.hidden_size) for _ in range(config.n_codes_total)]
+                ),
+                "wpe": nn.Embedding(config.block_size, config.hidden_size),
+                "drop": nn.Dropout(config.dropout),
+                "h": nn.ModuleList(
+                    [BarkBlock(config, layer_idx, is_causal=False) for layer_idx in range(config.num_layers)]
+                ),
+                "ln_f": nn.LayerNorm(config.hidden_size),
+            }
+        )
+        self.lm_heads = nn.ModuleList(
+            [
+                nn.Linear(config.hidden_size, config.output_vocab_size, bias=False)
+                for _ in range(config.n_codes_given, config.n_codes_total)
+            ]
+        )
+        for i in range(config.n_codes_total - config.n_codes_given):
+            self.transformer.wtes[i + 1].weight = self.lm_heads[i].weight
+
+    def get_num_params(self, non_embedding=True):
+        """
+        Return the number of parameters in the model. For non-embedding count (default), the position embeddings get
+        subtracted. The token embeddings would too, except due to the parameter sharing these params are actually used
+        as weights in the final layer, so we include them.
+        """
+        n_params = sum(p.numel() for p in self.parameters())
+        if non_embedding:
+            for wte in self.transformer.wtes:
+                n_params -= wte.weight.numel()
+            n_params -= self.transformer.wpe.weight.numel()
+        return n_params
+
+    def _get_and_check_input_embeddings(self, input_ids, input_embeds, pred_idx):
+        # the input_embeddings are the sum of the j previous codebooks embeddings before the current pred_idx codebook
+
+        if input_ids is not None and input_embeds is not None:
+            raise ValueError("You cannot specify both input_ids and input_embeds at the same time")
+        elif input_ids is not None:
+            batch_size, seq_len, codes = input_ids.size()
+            assert (
+                batch_size <= self.config.block_size
+            ), f"Cannot forward sequence of length {seq_len}, block size is only {self.config.block_size}"
+
+            assert codes == self.n_codes_total, (batch_size, seq_len, codes)
+
+            # forward the GPT model itself
+            input_embeds = [
+                wte(input_ids[:, :, i]).unsqueeze(-1) for i, wte in enumerate(self.transformer.wtes)
```

Comment:
I wonder if it might related to the training code, otherwise the number of embedded codebooks  summed changes w.r.t to an input variable. What do you think?
I leave it like this for now.

Reply:
Possibly, but since we don't return `input_embeds` to the user it's okay if we change how this is computed in the future (there won't be a breaking change)

I would advocate for doing it the cheaper way for now since we can always update it later to sum over all codebooks then just slice as many as we need

The cheaper way being:
```python
input_embeds = sum([self.transformer.wtes[i](input_ids[:, :, i, None]) for i in range(pred_idx + 1)])
```